File: 0966-binary-subarrays-with-sum/0966-binary-subarrays-with-sum.cpp
class Solution {
public:
    int numSubarraysWithSum(vector<int>& nums, int goal) {
        unordered_map<int, int> mapp;
        mapp[0] = 1;
        int sum = 0;
        int count = 0;

        for(int i = 0; i < nums.size(); i++){
            sum += nums[i];
            int remain = sum - goal;
            if(mapp.find(remain) != mapp.end()){
                count+= mapp[remain];
            }
            mapp[sum]++;
        }
        return count;
    }
};

File: 0966-binary-subarrays-with-sum/0966-binary-subarrays-with-sum.py
class Solution:
    def numSubarraysWithSum(self, nums: List[int], goal: int) -> int:
        summ = 0
        mapp = defaultdict(int)
        mapp[0] = 1 
        count = 0

        for i in range(len(nums)):
            summ += nums[i]
            remain = summ - goal
            if remain in mapp:
                count += mapp[remain]
            
            mapp[summ] = mapp[summ] + 1
        
        return count